reco_utils/recommender/tfidf/tfidf_utils.py

```diff
@@ -0,0 +1,400 @@
+        self.id_col = id_col
+        self.title_col = title_col
+        self.k = k
+
+        if self.k < 1:
+            raise ValueError("k cannot be < 1")
+
+        # Initialize other variables used in this class
+        self.tf = TfidfVectorizer()
+        self.tfidf_matrix = dict()
+        self.tokens = dict()
+        self.stop_words = frozenset()
+        self.recommendations = dict()
+        self.top_k_recommendations = pd.DataFrame()
+        
+    def __clean_text(self, text, for_BERT=False, verbose=False):
+        """ Clean text by removing HTML tags, symbols, and punctuation.
+        
+        Args:
+            text (str): Text to clean.
+            for_BERT (boolean): True or False for if this text is being cleaned for a BERT word tokenization method.
+            verbose (boolean): True or False for whether to print.
+        
+        Returns:
+            clean (str): Cleaned version of text.
+        """
+        
+        try:
+            # Normalize unicode
+            text_norm = unicodedata.normalize('NFC', text)
+
+            # Remove HTML tags
+            clean = re.sub('<.*?>', '', text_norm)
+
+            # Remove new line and tabs
+            clean = clean.replace('\n', ' ')
+            clean = clean.replace('\t', ' ')
+            clean = clean.replace('\r', ' ')
+            clean = clean.replace('Â\xa0', '')     # non-breaking space
+
+            # Remove all punctuation and special characters
+            clean = re.sub('([^\s\w]|_)+','', clean)
+
+            # If you want to keep some punctuation, see below commented out example
+            # clean = re.sub('([^\s\w\-\_\(\)]|_)+','', clean)
+
+            # Skip further processing if the text will be used in BERT tokenization
+            if for_BERT is False:
+                # Lower case
+                clean = clean.lower()
+        except:
+            if verbose is True:
+                print('Cannot clean non-existent text')
+            clean = ''
+        
+        return clean
+
+    def clean_dataframe(self, df, cols_to_clean, new_col_name='cleaned_text'):
+        """ Clean the text within the columns of interest and return a dataframe with cleaned and combined text.
+        
+        Args:
+            df (pd.DataFrame): Dataframe containing the text content to clean.
+            cols_to_clean (str): List of columns to clean by name (e.g., ['abstract','full_text']).
+            new_col_name (str): Name of the new column that will contain the cleaned text.
+
+        Returns:
+            df (pd.DataFrame): Dataframe with cleaned text in the new column.
+        """
+        # Check if for BERT tokenization
+        if self.tokenization_method in ['bert','scibert']:
+            for_BERT = True
+        else:
+            for_BERT = False
+
+        # Clean the text in the dataframe
+        for i in range(0, len(df)):
+            for col in cols_to_clean:
+                df[col][i] = self.__clean_text(df[col][i], for_BERT)
```

Comment:
Yes. Because this now occurs after joining `cols_to_clean` into a single column, I have opted for using `.map()` instead of `.apply()` on the single combined column. Changes have been made in the latest commit.